diff -uNr linux-3.18/arch/x86/boot/cpu.c linux-3.18.3-ossmga/arch/x86/boot/cpu.c
--- linux-3.18/arch/x86/boot/cpu.c	2014-12-07 22:21:05.000000000 +0000
+++ linux-3.18.3-ossmga/arch/x86/boot/cpu.c	2015-01-18 13:14:15.968113636 +0000
@@ -88,7 +88,7 @@
 	}
 
 	if (err_flags) {
-		puts("This kernel requires the following features "
+		puts("This boot kernel check detects requirement for following features "
 		     "not present on the CPU:\n");
 		show_cap_strs(err_flags);
 		putchar('\n');
diff -uNr linux-3.18/arch/x86/boot/main.c linux-3.18.3-ossmga/arch/x86/boot/main.c
--- linux-3.18/arch/x86/boot/main.c	2014-12-07 22:21:05.000000000 +0000
+++ linux-3.18.3-ossmga/arch/x86/boot/main.c	2015-01-18 13:16:21.336110408 +0000
@@ -147,9 +147,10 @@
 
 	/* Make sure we have all the proper CPU support */
 	if (validate_cpu()) {
-		puts("Unable to boot - please use a kernel appropriate "
-		     "for your CPU.\n");
-		die();
+		puts("If you did not use special compilation parameters, this is\n"
+		     "Unable to boot - please use a kernel appropriate "
+		     "for your CPU,\n in case of any future kernel crash!!!");
+		/*die();*/
 	}
 
 	/* Tell the BIOS what CPU mode we intend to run in. */
