--- linux/arch/x86/kernel/tsc.c	2015-01-29 19:41:03.000000000 -0600
+++ linux/arch/x86/kernel/tsc.c	2015-01-31 13:41:37.753496318 -0600
@@ -617,7 +617,7 @@
 			goto success;
 		}
 	}
-	pr_info("Fast TSC calibration failed\n");
+
 	return 0;
 
 success:
@@ -771,7 +771,7 @@
 
 		/* The alternative source failed as well, disable TSC */
 		if (tsc_ref_min == ULONG_MAX) {
-			pr_warn("HPET/PMTIMER calibration failed\n");
+
 			return 0;
 		}
 
@@ -790,7 +790,7 @@
 
 	/* The alternative source failed, use the PIT calibration value */
 	if (tsc_ref_min == ULONG_MAX) {
-		pr_warn("HPET/PMTIMER calibration failed. Using PIT calibration.\n");
+
 		return tsc_pit_min;
 	}
 
