--- linux/drivers/acpi/acpica/psparse.c	2015-01-29 19:41:03.000000000 -0600
+++ linux/drivers/acpi/acpica/psparse.c	2015-01-31 14:56:49.506702734 -0600
@@ -531,7 +531,7 @@
 
 			/* Either the method parse or actual execution failed */
 
-			ACPI_ERROR_METHOD("Method parse/execution failed",
+			ACPI_ERROR_METHOD("Method parse/execution info",
 					  walk_state->method_node, NULL,
 					  status);
 
