diff --git a/drivers/mfd/lpc_ich.c b/drivers/mfd/lpc_ich.c
index 027cc8f..7cac95b 100644
--- a/drivers/mfd/lpc_ich.c
+++ b/drivers/mfd/lpc_ich.c
@@ -704,7 +704,8 @@ static int __devinit lpc_ich_init_gpio(struct pci_dev *dev,
 	res = &gpio_ich_res[ICH_RES_GPE0];
 	res->start = base_addr + ACPIBASE_GPE_OFF;
 	res->end = base_addr + ACPIBASE_GPE_END;
-	ret = acpi_check_resource_conflict(res);
+	//ret = acpi_check_resource_conflict(res);
+	ret = 0;
 	if (ret) {
 		/*
 		 * This isn't fatal for the GPIO, but we have to make sure that
