diff -uNr a/Makefile b/Makefile
--- a/Makefile	2014-11-17 09:48:07.964034061 +0000
+++ b/Makefile	2014-11-17 09:47:31.977036601 +0000
@@ -1,8 +1,8 @@
 VERSION = 3
 PATCHLEVEL = 18
 SUBLEVEL = 10
-EXTRAVERSION =
-NAME = Diseased Newt
+EXTRAVERSION = .yainituall
+NAME = YaIniT with all ftom Gentoo Mageia OpenSuSE Ubuntu Fedora PCLOS and ZEN BFS BFQ MPTCP BLD UKSM
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
