diff -uNr a/Makefile b/Makefile
--- a/Makefile	2014-11-17 09:48:07.964034061 +0000
+++ b/Makefile	2014-11-17 09:47:31.977036601 +0000
@@ -1,8 +1,8 @@
 VERSION = 4
 PATCHLEVEL = 3
 SUBLEVEL = 3
-EXTRAVERSION =
-NAME = Blurry Fish Butt
+EXTRAVERSION = .yainitiuall_1601IV
+NAME = YaIniT with all from OpenSuSE-TW-9.1 IntelClear Gentoo Ubuntu-Xenial-7.18 Mageia6 sunxi AUR BFQ ZEN5 PF4
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
