diff -uNr a/Makefile b/Makefile
--- a/Makefile	2014-11-17 09:48:07.964034061 +0000
+++ b/Makefile	2014-11-17 09:47:31.977036601 +0000
@@ -1,8 +1,8 @@
 VERSION = 4
 PATCHLEVEL = 3
 SUBLEVEL = 3
-EXTRAVERSION =
-NAME = Blurry Fish Butt
+EXTRAVERSION = .yainitignuall_1601I
+NAME = YaIniT with all from OpenSuSE-4.1/4.3.3-3.1 Gentoo Ubuntu-Xenial-5.16 Mageia sunxi BFQ GNU-PCK2
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
