diff -uNr linux-3.19.3-pup/include/linux/jiffies.h linux-3.19.3-zen4yainit/include/linux/jiffies.h
--- linux-3.19.3-pup/include/linux/jiffies.h	2015-02-09 02:54:22.000000000 +0000
+++ linux-3.19.3-zen4yainit/include/linux/jiffies.h	2015-04-26 18:08:38.769405150 +0000
@@ -164,7 +164,11 @@ static inline u64 get_jiffies_64(void)
  * Have the 32 bit jiffies value wrap 5 minutes after boot
  * so jiffies wrap bugs show up earlier.
  */
+#if  defined(CONFIG_SCHED_BFS) || defined(CONFIG_PCK_INTERACTIVE)
+#define INITIAL_JIFFIES ((unsigned long)(unsigned int) (-10*HZ))
+#else
 #define INITIAL_JIFFIES ((unsigned long)(unsigned int) (-300*HZ))
+#endif
 
 /*
  * Change timeval to jiffies, trying to avoid the
