diff -uNr a/arch/x86/boot/cpu.c b/arch/x86/boot/cpu.c
--- a/arch/x86/boot/cpu.c	2014-12-07 22:21:05.000000000 +0000
+++ b/arch/x86/boot/cpu.c	2015-01-18 13:14:15.968113636 +0000
@@ -88,7 +88,7 @@
 	}
 
 	if (err_flags) {
-		puts("This kernel requires the following features "
-		     "not present on the CPU:\n");
+		puts("This kernel check detects requirement for following features "
+		     "not present on the running Processors:\n");
 		show_cap_strs(err_flags);
 		putchar('\n');
diff -uNr a/arch/x86/boot/main.c b/arch/x86/boot/main.c
--- a/arch/x86/boot/main.c	2014-12-07 22:21:05.000000000 +0000
+++ b/arch/x86/boot/main.c	2015-01-18 13:16:21.336110408 +0000
@@ -147,9 +147,14 @@
 
 	/* Make sure we have all the proper CPU support */
 	if (validate_cpu()) {
-		puts("Unable to boot - please use a kernel appropriate "
-		     "for your CPU.\n");
-		die();
+		puts("If you do not use compile configuration\n"
+		     "different from YaIniT,  this means:\n"
+		     "Unable to boot - \n"
+		     "please use a kernel appropriate for your CPU\n"
+		     "             ON FUTURE FAILURES ");
+		puts("Otherwise, this kernel will crash on any use\n"
+		     " of the detected missing feature!");
+		/*die();*/
 	}
 
 	/* Tell the BIOS what CPU mode we intend to run in. */
