diff --git a/net/ethernet/eth.c b/net/ethernet/eth.c
index 33a140e..8fcdda9 100644
--- a/net/ethernet/eth.c
+++ b/net/ethernet/eth.c
@@ -387,7 +387,7 @@ void ether_setup(struct net_device *dev)
 	dev->hard_header_len 	= ETH_HLEN;
 	dev->mtu		= ETH_DATA_LEN;
 	dev->addr_len		= ETH_ALEN;
-	dev->tx_queue_len	= 1000;	/* Ethernet wants good queues */
+	dev->tx_queue_len	= 256;	/* Ethernet wants good latency and queue. */
 	dev->flags		= IFF_BROADCAST|IFF_MULTICAST;
 	dev->priv_flags		|= IFF_TX_SKB_SHARING;
 
