diff -uprNw -X linux-4.1.vanilla/Documentation/dontdiff linux-4.1.vanilla/arch/x86/kernel/smpboot.c linux-4.1.missing-sched-domains/arch/x86/kernel/smpboot.c
--- linux-4.1.vanilla/arch/x86/kernel/smpboot.c	2015-06-21 22:05:43.000000000 -0700
+++ linux-4.1.missing-sched-domains/arch/x86/kernel/smpboot.c	2016-04-30 15:24:41.796729160 -0700
@@ -109,6 +109,10 @@
 /* Maximum number of SMT threads on any online core */
 int __max_smt_threads __read_mostly;
 
+#ifdef CONFIG_NUMA
+extern int sched_max_numa_distance;
+#endif
+
 static inline void smpboot_setup_warm_reset_vector(unsigned long start_eip)
 {
 	unsigned long flags;
@@ -558,7 +562,12 @@
 			} else if (i != cpu && !c->booted_cores)
 				c->booted_cores = cpu_data(i).booted_cores;
 		}
+#ifndef CONFIG_NUMA
 		if (match_die(c, o) && !topology_same_node(c, o))
+#else
+        if (match_die(c, o) && !topology_same_node(c, o)
+            && sched_max_numa_distance == -1)
+#endif
 			x86_has_numa_in_package = true;
 	}
 
diff -uprNw -X linux-4.1.vanilla/Documentation/dontdiff linux-4.1.vanilla/kernel/sched/core.c linux-4.1.missing-sched-domains/kernel/sched/core.c
--- linux-4.1.vanilla/kernel/sched/core.c	2015-06-21 22:05:43.000000000 -0700
+++ linux-4.1.missing-sched-domains/kernel/sched/core.c	2016-04-30 14:35:33.896865093 -0700
@@ -6415,7 +6415,7 @@
 static int sched_domains_numa_levels;
 enum numa_topology_type sched_numa_topology_type;
 static int *sched_domains_numa_distance;
-int sched_max_numa_distance;
+int sched_max_numa_distance = -1;
 static struct cpumask ***sched_domains_numa_masks;
 static int sched_domains_curr_level;
 #endif
