diff -uNr linux-3.16.7/net/ethernet/eth.c linux-3.16.7-lqx/net/ethernet/eth.c
--- linux-3.16.7/net/ethernet/eth.c	2014-08-03 22:25:02.000000000 +0000
+++ linux-3.16.7-lqx/net/ethernet/eth.c	2014-11-05 14:34:31.010077770 +0000
@@ -363,7 +363,7 @@
 	dev->hard_header_len 	= ETH_HLEN;
 	dev->mtu		= ETH_DATA_LEN;
 	dev->addr_len		= ETH_ALEN;
-	dev->tx_queue_len	= 1000;	/* Ethernet wants good queues */
+	dev->tx_queue_len	= 256;
 	dev->flags		= IFF_BROADCAST|IFF_MULTICAST;
 	dev->priv_flags		|= IFF_TX_SKB_SHARING;
 
