diff -uNr linux-3.16/drivers/video/logo/logo.c linux-3.16-utopic/drivers/video/logo/logo.c
--- linux-3.16/drivers/video/logo/logo.c	2014-08-03 22:25:02.000000000 +0000
+++ linux-3.16-utopic/drivers/video/logo/logo.c	2015-04-18 11:59:37.379266447 +0000
@@ -23,6 +23,21 @@
 #ifdef CONFIG_M68K
 #include <asm/setup.h>
 #endif
+/*
+ * Logos are located in the initdata, and will be freed in kernel_init.
+ * Use late_init to mark the logos as freed to prevent any further use.
+ */
+
+static bool logos_freed;
+
+static int __init fb_logo_late_init(void)
+{
+	logos_freed = true;
+	return 0;
+}
+
+late_initcall(fb_logo_late_init);
+
 
 static bool nologo;
 module_param(nologo, bool, 0);
@@ -134,7 +149,7 @@
 	const struct linux_logo **array = NULL;
 	unsigned int size;
  
-	if (nologo)
+	if (nologo || logos_freed)
  		return NULL;
  
 	/* Select logo array */
