diff -iwBEr -U 4 linux-3.16-ally/sound/pci/hda/hda_controller.c linux-3.16-37.49/sound/pci/hda/hda_controller.c
--- linux-3.16-ally/sound/pci/hda/hda_controller.c	2015-05-08 20:05:01.855023559 +0000
+++ linux-3.16-37.49/sound/pci/hda/hda_controller.c	2015-05-08 20:07:42.452040312 +0000
@@ -22,8 +22,9 @@
 
 #include <linux/clocksource.h>
 #include <linux/delay.h>
 #include <linux/interrupt.h>
+#include <linux/io.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/pm_runtime.h>
 #include <linux/slab.h>
diff -iwBEr -U 4 linux-3.16-ally/sound/pci/hda/patch_realtek.c linux-3.16-37.49/sound/pci/hda/patch_realtek.c
--- linux-3.16-ally/sound/pci/hda/patch_realtek.c	2015-05-08 20:05:01.933023567 +0000
+++ linux-3.16-37.49/sound/pci/hda/patch_realtek.c	2015-05-08 20:07:42.669040335 +0000
@@ -4815,8 +4815,11 @@
 	SND_PCI_QUIRK(0x1028, 0x06c7, "Dell", ALC255_FIXUP_DELL1_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x0668, "Dell", ALC255_FIXUP_DELL2_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x0669, "Dell", ALC255_FIXUP_DELL2_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x0684, "Dell", ALC269_FIXUP_DELL2_MIC_NO_PRESENCE),
+	SND_PCI_QUIRK(0x1028, 0x06c7, "Dell", ALC255_FIXUP_DELL1_MIC_NO_PRESENCE),
+	SND_PCI_QUIRK(0x1028, 0x06d9, "Dell", ALC293_FIXUP_DELL1_MIC_NO_PRESENCE),
+	SND_PCI_QUIRK(0x1028, 0x06da, "Dell", ALC293_FIXUP_DELL1_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x15cc, "Dell X5 Precision", ALC269_FIXUP_DELL2_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x15cd, "Dell X5 Precision", ALC269_FIXUP_DELL2_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x06d9, "Dell", ALC293_FIXUP_DELL1_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x06da, "Dell", ALC293_FIXUP_DELL1_MIC_NO_PRESENCE),
diff -iwBEr -U 4 linux-3.16-ally/tools/perf/builtin-timechart.c linux-3.16-37.49/tools/perf/builtin-timechart.c
--- linux-3.16-ally/tools/perf/builtin-timechart.c	2014-08-03 22:25:02.000000000 +0000
+++ linux-3.16-37.49/tools/perf/builtin-timechart.c	2015-05-08 20:07:42.452040312 +0000
@@ -739,10 +739,17 @@
 	/*
 	 * two pass drawing so that the P state bars are on top of the C state blocks
 	 */
 	while (pwr) {
-		if (pwr->type == CSTATE)
+		if (pwr->type == CSTATE) {
+			/* If the first event is an _end event, start timestamp is zero
+			   -> ignore these */
+			if (pwr->start_time == 0 || pwr->end_time == 0) {
+				pwr = pwr->next;
+				continue;
+			}
 			svg_cstate(pwr->cpu, pwr->start_time, pwr->end_time, pwr->state);
+		}
 		pwr = pwr->next;
 	}
 
 	pwr = tchart->power_events;
diff -iwBEr -U 4 linux-3.16-ally/include/linux/pci_ids.h linux-3.16-37.49/include/linux/pci_ids.h
--- linux-3.16-ally/include/linux/pci_ids.h	2015-05-08 20:05:01.232023494 +0000
+++ linux-3.16-37.49/include/linux/pci_ids.h	2015-05-08 20:07:42.415040308 +0000
@@ -74,8 +74,9 @@
 #define PCI_CLASS_SYSTEM_TIMER		0x0802
 #define PCI_CLASS_SYSTEM_RTC		0x0803
 #define PCI_CLASS_SYSTEM_PCI_HOTPLUG	0x0804
 #define PCI_CLASS_SYSTEM_SDHCI		0x0805
+#define PCI_CLASS_SYSTEM_IOMMU		0x0806
 #define PCI_CLASS_SYSTEM_OTHER		0x0880
 
 #define PCI_BASE_CLASS_INPUT		0x09
 #define PCI_CLASS_INPUT_KEYBOARD	0x0900
