--- a/arch/x86/platform/intel-mid/early_printk_intel_mid.c	2014-11-24 15:51:31.910111484 +0000
+++ b/arch/x86/platform/intel-mid/early_printk_intel_mid.c	2014-11-24 15:52:20.486106892 +0000
@@ -118,7 +118,7 @@
 	mrst_early_console_init();
 
 	while (kmsg_dump_get_line(dumper, true, line, sizeof(line), &len))
-		early_mrst_console.write(&early_mrst_console, line, len);
+		early_mrst_console.write(&early_mrst_console, line, sizeof(line), len);
 }
 
 /* Set the ratio rate to 115200, 8n1, IRQ disabled */
