diff -uNr linux-3.16-alluwoxen/arch/x86/boot/tools/build.c linux-3.16-allu/arch/x86/boot/tools/build.c
--- linux-3.16-alluwoxen/arch/x86/boot/tools/build.c	2014-08-03 22:25:02.000000000 +0000
+++ linux-3.16-allu/arch/x86/boot/tools/build.c	2015-04-19 21:04:22.703952739 +0000
@@ -356,8 +358,8 @@
 	c = fread(buf, 1, sizeof(buf), file);
 	if (ferror(file))
 		die("read-error on `setup'");
-	if (c < 1024)
-		die("The setup must be at least 1024 bytes");
+	if (c <= 512)
+		die("The setup must be more than 512 bytes");
 	if (get_unaligned_le16(&buf[510]) != 0xAA55)
 		die("Boot block hasn't got boot flag (0xAA55)");
 	fclose(file);
